### PR TITLE
Improve correlated subquery handling in ORCA

### PR DESCRIPTION
**Pull request by hardikar**
Previously, ORCA would produces wrong results for some queries with
outer references because of incorrect placement of Spool operator above
the operator with outer references. Sometimes ORCA would error out due
to illegal rescans, because a Spool was missing above a Motion on the
inner side of a correlated join. Also in some cases, ORCA would fall
back due to property mismatch. This commit fixes many of the above 
issues by refactoring RewindabilitySpec.

See each individual commit for the changes & notes regarding tests.

PS: In an attempt to open a PR earlier, we decided to add TODOs in the
codebase. These pending issues need more thought and/or they represent
existing bugs discovered while working on the current PR.

GPDB side PR: https://github.com/greenplum-db/gpdb/pull/7078

### CODE REVIEW

File: libgpopt/src/operators/CPhysicalFilter.cpp [hunk 1]
```diff
@@ -166,16 +166,6 @@ CPhysicalFilter::PrsRequired
...
-	// child), we can optimize by adding a materialize in between. However, if
-	// there are outer references in the child, we should *not* add a materialize
-	// here.  Otherwise the child will not get rescanned leading to wrong
-	// results.
-	if (exprhdl.HasOuterRefs() && !exprhdl.HasOuterRefs(0))
-	{
-		return GPOS_NEW(mp) CRewindabilitySpec(CRewindabilitySpec::ErtRewindable, prsRequired->Emht());
-	}
-
 	return PrsPassThru(mp, exprhdl, prsRequired, child_index);
```
Comment (zellerh) [inline libgpopt/src/operators/CPhysicalFilter.cpp]: Same question as above: Is there something special to do if prsRequired asks for rewindability and if the filter predicate contains an outer reference?

File: data/dxl/minidump/MultipleSubqueriesInSelectClause.mdp [hunk 2]
```diff
@@ -1174,7 +1187,7 @@
     <dxl:Plan Id="0" SpaceSize="1441">
       <dxl:Result>
         <dxl:Properties>
-          <dxl:Cost StartupCost="0" TotalCost="9309.072976" Rows="3000000.000000" Width="32"/>
+          <dxl:Cost StartupCost="0" TotalCost="9309.072916" Rows="3000000.000000" Width="32"/>
```
Comment (bhuvnesh2703) [inline data/dxl/minidump/MultipleSubqueriesInSelectClause.mdp]: i was checking how come without any data the rows estimate is 300...., so i see the table `test_fct_unnest` which is there in the dxl tree but is it missing from the DDL above and the data?

File: data/dxl/minidump/NestedProjectCountStarWithOuterRefs.mdp [hunk 3]
```diff
@@ -1,5 +1,11 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <dxl:DXLMessage xmlns:dxl="http://greenplum.com/dxl/2010/12/">
+ <dxl:Comment><![CDATA[
+  create table x(i int, j int);
+  create table y(i int, j int);
```
Comment (bhuvnesh2703) [inline data/dxl/minidump/NestedProjectCountStarWithOuterRefs.mdp]: insert statement.. ? x and y both has data.

File: libgpopt/include/gpopt/base/CRewindabilitySpec.h [hunk 4]
```diff
@@ -35,45 +35,68 @@ namespace gpopt
...
-			// 2. RewindableNoMotion - {ErtRewindable, EmhtNoMotion}
-			//	  require: I require my child to be rewindable and also motion hazard handling is unnecessary.
-			//	  derive: I am rewindable and I do not impose motion hazard (any rewindable operator without
-			//	  a motion underneath it or a blocking spool with a motion underneath it, will derive this)
+			// 1. ErtRewindable:
+			//    require:
+			//      I require my child to be rewindable. This is required by NLJ (for
+			//      example) of its inner child.
+			//    derive:
+			//      I am rewindable. (e.g Spool, Sort, Scan)
```
[the remark this answers is not in the document]
Comment (bhuvnesh2703) [reply libgpopt/include/gpopt/base/CRewindabilitySpec.h]: @zellerh If the underlying tree has an outer reference, it should not be rewindable but rescannable atleast.. 
So i believe, rewindable is not allowed in such case as it may give wrong results. 
@hardikar @chrishajas is it correct?

File: data/dxl/minidump/MultipleSubqueriesInSelectClause.mdp (hunk 2 above)
Comment (chrishajas) [reply data/dxl/minidump/MultipleSubqueriesInSelectClause.mdp]: Oh good point. The place where I found this query doesn't actually match this query/plan, so we should remove the ddl/query from this mdp.

File: data/dxl/minidump/MultipleSubqueriesInSelectClause.mdp [hunk 5]
```diff
@@ -1,5 +1,18 @@
...
+  create table t1 (c1 integer);
+  create table t2 (c2 integer);
+  create table t3 (c3 integer);
+  create table t4 (c4 integer);
+
+  select c1 from t1 where c1 not in
+  (select c2 from t2 where c2 not in
+  (select c3 from t3 where c3 not in
+  (select c4 from t4 where c4 = c2 )));
+
```
Comment (chrishajas) [inline data/dxl/minidump/MultipleSubqueriesInSelectClause.mdp]: todo: remove this comment. The ddl/query I found doesn't actually match this minidump

File: data/dxl/minidump/NestedProjectCountStarWithOuterRefs.mdp [hunk 6]
```diff
@@ -1,5 +1,11 @@
...
+  explain select (select 1+ (select 1+count(*) from x where y.j=x.j) from (values(1))r ) from y;
```
Comment (chrishajas) [inline data/dxl/minidump/NestedProjectCountStarWithOuterRefs.mdp]: todo: remove this comment. I can't find the data used to generate this minidump and I think the query is actually wrong too. It'd be best to just remove this comment too.

File: libgpopt/include/gpopt/base/CRewindabilitySpec.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (hardikar) [reply libgpopt/include/gpopt/base/CRewindabilitySpec.h]: @bhuvnesh2703 That's right.

File: libgpopt/src/operators/CPhysicalFilter.cpp (hunk 1 above)
Comment (hardikar) [reply libgpopt/src/operators/CPhysicalFilter.cpp]: I don't think there is anything special to do in this case. If the filter had an outer ref, there would have been a parent that asked for Rescannable. However, an intermediate op could "upgrade" that to a Rewindable. If that happens, and the child of the filter is not rewindable, a Material will be added above the child. And if the child is rewindable, then we will add a Material where ever need above the Filter. In the executor a Material node behaves appropriately based on if there is param change or not. If there is a param change, it throws out the materialized results; if there is no param change, it simple resets the iterator.

In summary, the presence of an outer ref in the Filter makes no difference to the Filter, i.e a Material on top of a Filter with outer refs is going to be fine.

File: libgpopt/include/gpopt/base/CRewindabilitySpec.h [hunk 7]
```diff
@@ -35,45 +35,68 @@ namespace gpopt
...
+			// following (in required & derived contexts):
 			//
-			// 2. RewindableNoMotion - {ErtRewindable, EmhtNoMotion}
-			//	  require: I require my child to be rewindable and also motion hazard handling is unnecessary.
-			//	  derive: I am rewindable and I do not impose motion hazard (any rewindable operator without
-			//	  a motion underneath it or a blocking spool with a motion underneath it, will derive this)
+			// 1. ErtRewindable:
+			//    require:
+			//      I require my child to be rewindable. This is required by NLJ (for
+			//      example) of its inner child.
```
[the remark this answers is not in the document]
Comment (hardikar) [reply libgpopt/include/gpopt/base/CRewindabilitySpec.h]: I'm just going to remove the for example if that's confusing. But yes NLJ requires its inner child to be rewindable.

File: libgpopt/include/gpopt/base/CRewindabilitySpec.h [hunk 8]
```diff
@@ -35,45 +35,68 @@ namespace gpopt
...
+			//      I am rewindable. (e.g Spool, Sort, Scan)
 			//
-			// 3. NotRewindableMotion - {ErtNotRewindable, EmhtMotion}
-			//	  require: I do not require my child to be rewindable but it may need to handle motion hazard.
-			//	  derive: I am not rewindable and I impose motion hazard (all motions except gather motion will
-			//    derive this)
+			// 2. ErtRescannable:
+			//    require:
+			//      I require my child to be rescannable, so that I can re-execute
+			//      the entire subtree if needed. This is required by correlated
```
[the remark this answers is not in the document]
Comment (hardikar) [reply libgpopt/include/gpopt/base/CRewindabilitySpec.h]: Ah I see, we're getting into language usage here. Require from vs require of. I think either work in this case, since in the comments each operator is kind of personified. I'll just change it to from to make it simpler.

File: libgpopt/include/gpopt/base/CRewindabilitySpec.h [hunk 9]
```diff
@@ -144,6 +167,16 @@ namespace gpopt
 				return Ert() == ErtRewindable;
 			}
 
+			BOOL IsRescannable() const
+			{
+				return Ert() == ErtRescannable;
+			}
+
+			BOOL IsCheckRequired() const
```
[the remark this answers is not in the document]
Comment (hardikar) [reply libgpopt/include/gpopt/base/CRewindabilitySpec.h]: I must protest this. The purpose of this renaming was a hope to later refactor all of the property specs to have a function called `IsCheckRequired`. It is used in many places, and we chased a bug for a couple of days because we forgot to modify something in one place and not the other.

If you want I can add *another* function called `IsRescannableOrRewindable` and have `IsCheckRequired` call that. But this function was added for a different purpose than just a not-None check.

File: data/dxl/minidump/NestedProjectCountStarWithOuterRefs.mdp (hunk 6 above)
Comment (hardikar) [reply data/dxl/minidump/NestedProjectCountStarWithOuterRefs.mdp]: @chrishajas Why do you think the query is wrong?

File: data/dxl/minidump/MultipleSubqueriesInSelectClause.mdp (hunk 5 above)
Comment (hardikar) [reply data/dxl/minidump/MultipleSubqueriesInSelectClause.mdp]: done

File: data/dxl/minidump/NestedProjectCountStarWithOuterRefs.mdp (hunk 6 above)
Comment (chrishajas) [reply data/dxl/minidump/NestedProjectCountStarWithOuterRefs.mdp]: Just checked, query looks fine. My memory was wrong. Also, for the stats, it appears that table x has 1:9000 for i, and 1:9000 for j. For table y, it has 1:10 for i, and 1:2 for j.

Comment (chrishajas) [reply data/dxl/minidump/NestedProjectCountStarWithOuterRefs.mdp]:
```
insert into x select i,i from generate_series(1,9011) i;
insert into y select i,i%2 from generate_series(1,10) i;
```
